File: packages/contracts/python/sros_contracts/research_context.py
"""ResearchContext: the research specification.

Ontology V2 section 11.3. This is a **value object**, not an entity: it has no
identity and no lifecycle, and two contexts with identical parameters are the
same specification.

It is serialized as an **immutable snapshot** on a ResearchSession. The snapshot
is the reproducibility guarantee: editing a project default context must never
retroactively change what a past session says it ran with.

``canonical_json()`` is what gets persisted and hashed. It is deterministic --
sorted keys, canonicalized members -- so the same specification always produces
the same bytes, which is what makes the snapshot comparable and its hash stable.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field, replace

from .errors import ContractError
from .generated.domain import RESEARCH_CONTEXT_SCHEMA_VERSION
from .market_scope import MarketScope
from .registry import RegistryRef

__all__ = ["ResearchContext", "BudgetConstraints", "RESEARCH_CONTEXT_SCHEMA_VERSION"]

_FIELDS = {
    "schema_version",
    "market_scope",
    "market_types",
    "product_types",
    "domains",
    "audience",
    "languages",
    "budget_constraints",
    "technical_constraints",
    "desired_mvp_complexity",
    "research_depth",
    "time_horizon_days",
    "excluded_markets",
    "excluded_categories",
    "filters",
}


@dataclass(frozen=True)
class BudgetConstraints:
    """Per-session cost ceiling. Concrete figures are configuration (ADR-006).

    Budget exhaustion is COMPLETED with reduced Research Completeness, never a
    session failure (Ontology V2 section 15).
    """

    max_cost_units: float | None = None
    max_llm_calls: int | None = None

    def __post_init__(self) -> None:
        if self.max_cost_units is not None and self.max_cost_units < 0:
            raise ContractError("budget_constraints.max_cost_units", "must not be negative")
        if self.max_llm_calls is not None and self.max_llm_calls < 0:
            raise ContractError("budget_constraints.max_llm_calls", "must not be negative")

    def to_json(self) -> dict[str, object]:
        return {"max_cost_units": self.max_cost_units, "max_llm_calls": self.max_llm_calls}

    @classmethod
    def from_json(cls, data: object) -> BudgetConstraints:
        if not isinstance(data, dict):
            raise ContractError("budget_constraints", "expected an object")
        unknown = set(data) - {"max_cost_units", "max_llm_calls"}
        if unknown:
            raise ContractError("budget_constraints", f"unknown fields: {sorted(unknown)}")
        return cls(
            max_cost_units=data.get("max_cost_units"),
            max_llm_calls=data.get("max_llm_calls"),
        )


@dataclass(frozen=True)
class ResearchContext:
    """Immutable research specification. Not an entity."""

    market_scope: MarketScope
    market_types: tuple[RegistryRef, ...] = ()
    product_types: tuple[RegistryRef, ...] = ()
    domains: tuple[str, ...] = ()
    audience: str | None = None
    languages: tuple[str, ...] = ()
    budget_constraints: BudgetConstraints | None = None
    technical_constraints: tuple[str, ...] = ()
    desired_mvp_complexity: str | None = None
    research_depth: str | None = None
    time_horizon_days: int | None = None
    excluded_markets: tuple[RegistryRef, ...] = ()
    excluded_categories: tuple[str, ...] = ()
    filters: dict[str, object] = field(default_factory=dict)

    schema_version: str = RESEARCH_CONTEXT_SCHEMA_VERSION

    def __post_init__(self) -> None:
        if not isinstance(self.market_scope, MarketScope):
            raise ContractError("research_context.market_scope", "a MarketScope is required")

        if self.time_horizon_days is not None and (
            not isinstance(self.time_horizon_days, int)
            or isinstance(self.time_horizon_days, bool)
            or self.time_horizon_days < 1
        ):
            raise ContractError("research_context.time_horizon_days", "must be a positive integer")

        normalized = tuple(sorted({lang.lower() for lang in self.languages}))
        object.__setattr__(self, "languages", normalized)

        for ref in (*self.market_types, *self.product_types, *self.excluded_markets):
            if not isinstance(ref, RegistryRef):
                raise ContractError("research_context", "registry fields require a RegistryRef")

    # -- serialization ------------------------------------------------------

    @classmethod
    def from_json(cls, data: object) -> ResearchContext:
        if not isinstance(data, dict):
            raise ContractError("research_context", "expected an object")

        unknown = set(data) - _FIELDS
        if unknown:
            raise ContractError(
                "research_context",
                f"unknown fields: {sorted(unknown)}. Unknown fields are rejected so a "
                "typo is not silently dropped from a snapshot.",
            )
        if "market_scope" not in data:
            raise ContractError("research_context.market_scope", "required")

        budget = data.get("budget_constraints")
        return cls(
            market_scope=MarketScope.from_json(data["market_scope"]),
            market_types=_refs(data.get("market_types")),
            product_types=_refs(data.get("product_types")),
            domains=tuple(data.get("domains") or ()),
            audience=data.get("audience"),
            languages=tuple(data.get("languages") or ()),
            budget_constraints=BudgetConstraints.from_json(budget) if budget else None,
            technical_constraints=tuple(data.get("technical_constraints") or ()),
            desired_mvp_complexity=data.get("desired_mvp_complexity"),
            research_depth=data.get("research_depth"),
            time_horizon_days=data.get("time_horizon_days"),
            excluded_markets=_refs(data.get("excluded_markets")),
            excluded_categories=tuple(data.get("excluded_categories") or ()),
            filters=dict(data.get("filters") or {}),
            schema_version=data.get("schema_version", RESEARCH_CONTEXT_SCHEMA_VERSION),
        )

    def to_json(self) -> dict[str, object]:
        payload: dict[str, object] = {
            "audience": self.audience,
            "desired_mvp_complexity": self.desired_mvp_complexity,
            "domains": list(self.domains),
            "excluded_categories": list(self.excluded_categories),
            "excluded_markets": [r.to_json() for r in self.excluded_markets],
            "filters": dict(self.filters),
            "languages": list(self.languages),
            "market_scope": self.market_scope.to_json(),
            "market_types": [r.to_json() for r in self.market_types],
            "product_types": [r.to_json() for r in self.product_types],
            "research_depth": self.research_depth,
            "schema_version": self.schema_version,
            "technical_constraints": list(self.technical_constraints),
            "time_horizon_days": self.time_horizon_days,
        }
        if self.budget_constraints is not None:
            payload["budget_constraints"] = self.budget_constraints.to_json()
        return payload

    def canonical_json(self) -> str:
        """Deterministic serialization. This is what gets snapshotted and hashed."""
        return json.dumps(self.to_json(), sort_keys=True, separators=(",", ":"), ensure_ascii=False)

    def snapshot_hash(self) -> str:
        """Stable content hash of the canonical form."""
        return hashlib.sha256(self.canonical_json().encode("utf-8")).hexdigest()

    def with_changes(self, **changes: object) -> ResearchContext:
        """Return a NEW context. The original is never mutated."""
        return replace(self, **changes)  # type: ignore[arg-type]


def _refs(value: object) -> tuple[RegistryRef, ...]:
    if value is None:
        return ()
    if not isinstance(value, (list, tuple)):
        raise ContractError("research_context", "registry fields expect a list")
    return tuple(
        item if isinstance(item, RegistryRef) else RegistryRef.from_json(item) for item in value
    )

File: packages/contracts/src/researchContext.ts
/**
 * ResearchContext: the research specification.
 *
 * Ontology V2 §11.3. A **value object**, not an entity: no identity, no
 * lifecycle, and two contexts with identical parameters are the same
 * specification.
 *
 * Serialized as an **immutable snapshot** on a ResearchSession. The snapshot is
 * the reproducibility guarantee: editing a project's default context must never
 * retroactively change what a past session says it ran with.
 *
 * `canonicalJson` is what gets persisted and hashed — deterministic, sorted
 * keys, canonicalized members — so the same specification always produces the
 * same bytes.
 */

import { ContractError } from "./errors.ts";
import { RESEARCH_CONTEXT_SCHEMA_VERSION } from "./generated/domain.ts";
import {
  marketScopeToJson,
  parseMarketScope,
  type MarketScope,
} from "./marketScope.ts";
import { parseRegistryRef, registryRefToJson, type RegistryRef } from "./registry.ts";

const FIELDS = new Set([
  "schema_version",
  "market_scope",
  "market_types",
  "product_types",
  "domains",
  "audience",
  "languages",
  "budget_constraints",
  "technical_constraints",
  "desired_mvp_complexity",
  "research_depth",
  "time_horizon_days",
  "excluded_markets",
  "excluded_categories",
  "filters",
]);

export interface BudgetConstraints {
  readonly max_cost_units: number | null;
  readonly max_llm_calls: number | null;
}

export interface ResearchContext {
  readonly schema_version: string;
  readonly market_scope: MarketScope;
  readonly market_types: readonly RegistryRef[];
  readonly product_types: readonly RegistryRef[];
  readonly domains: readonly string[];
  readonly audience: string | null;
  readonly languages: readonly string[];
  readonly budget_constraints: BudgetConstraints | null;
  readonly technical_constraints: readonly string[];
  readonly desired_mvp_complexity: string | null;
  readonly research_depth: string | null;
  readonly time_horizon_days: number | null;
  readonly excluded_markets: readonly RegistryRef[];
  readonly excluded_categories: readonly string[];
  readonly filters: Readonly<Record<string, unknown>>;
}

function strings(value: unknown, field: string): string[] {
  if (value === undefined || value === null) return [];
  if (!Array.isArray(value)) throw new ContractError(field, "expected a list");
  return value.map((item) => {
    if (typeof item !== "string") throw new ContractError(field, "expected strings");
    return item;
  });
}

function refs(value: unknown): RegistryRef[] {
  if (value === undefined || value === null) return [];
  if (!Array.isArray(value)) {
    throw new ContractError("research_context", "registry fields expect a list");
  }
  return value.map(parseRegistryRef);
}

function parseBudget(value: unknown): BudgetConstraints | null {
  if (value === undefined || value === null) return null;
  if (typeof value !== "object" || Array.isArray(value)) {
    throw new ContractError("budget_constraints", "expected an object");
  }
  const record = value as Record<string, unknown>;
  const unknown = Object.keys(record).filter(
    (k) => k !== "max_cost_units" && k !== "max_llm_calls",
  );
  if (unknown.length > 0) {
    throw new ContractError("budget_constraints", `unknown fields: ${unknown.sort().join(", ")}`);
  }
  const cost = record["max_cost_units"] ?? null;
  const calls = record["max_llm_calls"] ?? null;
  if (cost !== null && (typeof cost !== "number" || cost < 0)) {
    throw new ContractError("budget_constraints.max_cost_units", "must not be negative");
  }
  if (calls !== null && (typeof calls !== "number" || calls < 0)) {
    throw new ContractError("budget_constraints.max_llm_calls", "must not be negative");
  }
  return Object.freeze({
    max_cost_units: cost,
    max_llm_calls: calls,
  });
}

export function parseResearchContext(data: unknown): ResearchContext {
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new ContractError("research_context", "expected an object");
  }
  const record = data as Record<string, unknown>;

  const unknown = Object.keys(record).filter((k) => !FIELDS.has(k));
  if (unknown.length > 0) {
    throw new ContractError(
      "research_context",
      `unknown fields: ${unknown.sort().join(", ")}. Unknown fields are rejected so a ` +
        `typo is not silently dropped from a snapshot.`,
    );
  }
  if (!("market_scope" in record)) {
    throw new ContractError("research_context.market_scope", "required");
  }

  const horizon = record["time_horizon_days"] ?? null;
  if (
    horizon !== null &&
    (typeof horizon !== "number" || !Number.isInteger(horizon) || horizon < 1)
  ) {
    throw new ContractError("research_context.time_horizon_days", "must be a positive integer");
  }

  const languages = [
    ...new Set(strings(record["languages"], "research_context.languages").map((l) => l.toLowerCase())),
  ].sort();

  const audience = record["audience"] ?? null;
  const mvp = record["desired_mvp_complexity"] ?? null;
  const depth = record["research_depth"] ?? null;

  return Object.freeze({
    schema_version:
      typeof record["schema_version"] === "string"
        ? record["schema_version"]
        : RESEARCH_CONTEXT_SCHEMA_VERSION,
    market_scope: parseMarketScope(record["market_scope"]),
    market_types: Object.freeze(refs(record["market_types"])),
    product_types: Object.freeze(refs(record["product_types"])),
    domains: Object.freeze(strings(record["domains"], "research_context.domains")),
    audience: typeof audience === "string" ? audience : null,
    languages: Object.freeze(languages),
    budget_constraints: parseBudget(record["budget_constraints"]),
    technical_constraints: Object.freeze(
      strings(record["technical_constraints"], "research_context.technical_constraints"),
    ),
    desired_mvp_complexity: typeof mvp === "string" ? mvp : null,
    research_depth: typeof depth === "string" ? depth : null,
    time_horizon_days: horizon,
    excluded_markets: Object.freeze(refs(record["excluded_markets"])),
    excluded_categories: Object.freeze(
      strings(record["excluded_categories"], "research_context.excluded_categories"),
    ),
    filters: Object.freeze({ ...(record["filters"] as Record<string, unknown> | undefined) }),
  });
}

type Json = Record<string, unknown>;

export function researchContextToJson(ctx: ResearchContext): Json {
  const payload: Json = {
    audience: ctx.audience,
    desired_mvp_complexity: ctx.desired_mvp_complexity,
    domains: [...ctx.domains],
    excluded_categories: [...ctx.excluded_categories],
    excluded_markets: ctx.excluded_markets.map(registryRefToJson),
    filters: { ...ctx.filters },
    languages: [...ctx.languages],
    market_scope: marketScopeToJson(ctx.market_scope),
    market_types: ctx.market_types.map(registryRefToJson),
    product_types: ctx.product_types.map(registryRefToJson),
    research_depth: ctx.research_depth,
    schema_version: ctx.schema_version,
    technical_constraints: [...ctx.technical_constraints],
    time_horizon_days: ctx.time_horizon_days,
  };
  if (ctx.budget_constraints !== null) {
    payload["budget_constraints"] = { ...ctx.budget_constraints };
  }
  return payload;
}

/** Deterministic JSON with recursively sorted keys. Mirrors Python `json.dumps(sort_keys=True)`. */
export function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    // JSON.stringify returns undefined for `undefined` and for functions.
    // The declared type hides that, so the fallback stays.
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    return JSON.stringify(value) ?? "null";
  }
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(",")}]`;
  const record = value as Record<string, unknown>;
  const parts = Object.keys(record)
    .sort()
    .map((key) => `${JSON.stringify(key)}:${stableStringify(record[key])}`);
  return `{${parts.join(",")}}`;
}

/** What gets persisted and hashed. Byte-identical to the Python implementation. */
export const researchContextCanonicalJson = (ctx: ResearchContext): string =>
  stableStringify(researchContextToJson(ctx));
